Reduce the use of `cp` in the testsuite

shindere opened the pull request:
This is a follow-up to issue #8659.

The present PR does two things:

 1. It renames ocamltest's `files` variable to `readonly_files`, to make it clear that files declared that way won't be modified by the test.
 2. It introduces and makes use of a `subdirectories` variable: it lists directories to be recursively copied from the test's source tree to its build tree.

Not all occurrences of `cp`could be eliminated. The remaining ones are of two kinds:

 1. From the source to the build tree but with a different name
 2. Between two locations within the test's build tree

I don't know whether such copies could be expressed in a more declarative way. I didn't try to modify these tests to make them fit the existing feature set.

shindere committed:
ocamltest: rename the files variable to readonly_files

shindere committed:
ocamltest: introduce the subdirectories variable

It lets one specify directories that should be recursively copied from
the test source directory to its build directory.

shindere committed:
Rewrite the lib-dynlink-native test to use ocamltest's subdirectories variable

shindere committed:
Rewrite the lib-dynlink-pr4229 test to use ocamltest's subdirectories variable

shindere committed:
Rewrite the lib-dynlink-pr4839 test to use ocamltest's subdirectories variable

shindere committed:
Slightly simplify the lib-dynlink-private test

Use relative paths to define the program variable and use more
homogeneous names for bytecode and native programs.

shindere committed:
Put lib-dynlink-private in a correct form

shindere committed:
Rewrite the lib-dynlink-private test to use ocamltest's subdirectories variable

shindere committed:
Rewrite the opaque test to use ocamltest's subdirectories variable

shindere committed:
Rewrite the missing_set_of_closures test to use ocamltest's subdirectories variable

shindere committed:
Rewrite the tool-ocamldep-shadowing test to use ocamltest's subdirectories variable

shindere committed:
Rewrite the typing-missing-cmi test to use ocamltest's subdirectories variable

dra27 commented on ocamltest/ocamltest_stdlib.ml:
```diff
@@ -173,6 +173,20 @@ module Sys = struct
     with_output_file ~bin:true dest @@ fun oc ->
     copy_chan ic oc
 
+  let rec copy_directory src dst =
```
This implementation can be tripped up by symlinks - I don't think it matters, but we could pull `Unix.stat` into `ocamltest_unix` (as the machinery is there for other reasons) and reject directories containing symlinks?

dra27 commented on testsuite/tests/tool-toplevel/error_highlighting.ml:
```diff
@@ -1,8 +1,8 @@
-            error_highlighting_use3.ml \
-            error_highlighting_use4.ml"
+   readonly_files = "error_highlighting_use1.ml \
+                    error_highlighting_use2.ml \
+                    error_highlighting_use3.ml \
+                    error_highlighting_use4.ml"
```
Micro-nit: these lines want an extra space!

dra27 reviewed (commented):
The simplification of those tests is definitely nice! The implementation looks fine to me, apart from possible hardening against symlinks?

I had a look at the remaining tests which still have `cp` in them:

- `no-alias-deps/aliases.ml`
	`b.cmi.invalid` could now be renamed to `b.cmi` and committed. It couldn't be called that in the old `Makefile`-system because it would be cleaned.

- `opaque/test.ml`
	The `cp` here includes items built by a previous step. This could be replaced by an action:
```
**** copy
src = "intf"
dst = "fst"
```

- `regression/missing_set_of_closures/missing_set_of_closures.ml`
	Likewise, this is copying items built by a previous step, so it could be a candidate for a new action.

- `tool-ocamldep-modalias/main.ml`
	There are two uses of cp here: the first is to `cp lib_impl.ml lib.ml`. That could use a copy action as with the previous ones (IIUC the point is that the environment needs to be set-up without that file being captured for dependencies).

	The second use replaces `Makefile` with `Makefile.build` and `Makefile.build2` - this seems unnecessary, why don't we just use `${MAKE} -f Makefile.build` and `${MAKE} -f Makefile.build2` later on?

- `tool-ocamlopt-save-ir/start_from_emit.ml`
	The `cp` again operates on an item in an earlier step, so it could also use a new copy action.

Adding a new action is probably beyond the scope of this PR (unless you want to), but I think it might be worth quickly fixing the `b.cmi.invalid` and `Makefile.build*` parts while here?

shindere committed:
Slightly simplify the tool-ocamldep-modalias test

shindere commented:
Thanks a lot for the feedback.

David Allsopp (2021/04/12 03:41 -0700):
> The simplification of those tests is definitely nice! The
> implementation looks fine to me, apart from possible hardening against
> symlinks?

Just one general comment: I do think that the simplification of the
tests was worth the effort and I am happy towork on the topic. Still, at
an abstract level, it's not so clear to me why having `cp` in tests is
so bad that we would want to eliminate each and every of its
occurrences?

> I had a look at the remaining tests which still have `cp` in them:
>
> - `no-alias-deps/aliases.ml`
> 	`b.cmi.invalid` could now be renamed to `b.cmi` and committed.
> 	It couldn't be called that in the old `Makefile`-system because
> 	it would be cleaned.

I was tempted to do that, yes, and still willing to. I was just a bit
puzzled to lose the "invalid" in th ename, which somehow looked
important to me. Shall I go for "B_invalid.cmi", for isntace?

> - `opaque/test.ml`
> 	The `cp` here includes items built by a previous step. This could be replaced by an action:
> ```
> **** copy
> src = "intf"
> dst = "fst"
> ```

So, in your opinion, what would be the semantics of this "copy" action,
exactly?
Would it be like a "cp" but between items in the test's build tree? It's
not completely clear relatively to which directory the paths should be
interpreted, actually.

> - `regression/missing_set_of_closures/missing_set_of_closures.ml`
> 	Likewise, this is copying items built by a previous step, so it could be a candidate for a new action.
>
> - `tool-ocamldep-modalias/main.ml`
> 	There are two uses of cp here: the first is to `cp lib_impl.ml lib.ml`. That could use a copy action as with the previous ones (IIUC the point is that the environment needs to be set-up without that file being captured for dependencies).
>
> 	The second use replaces `Makefile` with `Makefile.build` and
> 	`Makefile.build2` - this seems unnecessary, why don't we just
> 	use `${MAKE} -f Makefile.build` and `${MAKE} -f Makefile.build2`
> 	later on?

I implemented the second part already, thanks!

> - `tool-ocamlopt-save-ir/start_from_emit.ml`
> 	The `cp` again operates on an item in an earlier step, so it could also use a new copy action.
>
> Adding a new action is probably beyond the scope of this PR (unless
> you want to), but I think it might be worth quickly fixing the
> `b.cmi.invalid` and `Makefile.build*` parts while here?

Yes. Just let me know what you think about the name I suggested and I'll
proceed. I admit that if you do'nt mind keeping "copy" for later, that
would be nice. But if this PR does not include the copy action, do
thinkwe can still close the original issue?

> > @@ -173,6 +173,20 @@ module Sys = struct
>      with_output_file ~bin:true dest @@ fun oc ->
>      copy_chan ic oc
>
> +  let rec copy_directory src dst =
>
> This implementation can be tripped up by symlinks - I don't think it
> matters, but we could pull `Unix.stat` into `ocamltest_unix` (as the
> machinery is there for other reasons) and reject directories
> containing symlinks?

I don't know. Recently I had a look to that and felt a bit uncomfortable
because I realised there may be the need to distinguish (1) the libunix
that is used by ocamltest and (2) the libunix used to link tsets
against. The whole thingfelt complicated to me and I wished ocamltest
didn't depend on libunix at all itself.

> +                    error_highlighting_use2.ml \
> +                    error_highlighting_use3.ml \
> +                    error_highlighting_use4.ml"
>
> Micro-nit: these lines want an extra space!

What's that, sorry?

dra27 commented:
For the general motivation, it's twofold: there is no `cp` command on Windows (there is native GNU make, though) and I prefer the ocamltest specification to do the "plumbing" directly (i.e. actions and settings specify how the environment is set-up rather than side-effects of commands). In the short-term, every Cygwin command we invoke from a native Windows application is another chance for Cygwin to fail a CI job! :o)

For `no-alias-deps/aliases.ml` - the important point for the test is that `b.cmi` needs to exist and it needs to be a corrupt file. So it does have to be called `b.cmi`. In the old `Makefile` days, that wouldn't have worked because `make -C testsuite clean` would have erased it, but we don't do that anymore. So it could just be committed to the repo as `b.cmi` (it either has to be done with `git add -f` or, pedantically, a `!` pattern can be put in `.gitignore` for it somewhere after that `*.cmi` pattern.

For both `opaque/test.ml` and `regression/missing_set_of_closures/missing_set_of_closures.ml`, having `copy` take `src` as a list of files and `dst` as a directory is what's needed. For `tool-ocamldep-modalias/main.ml` and `tool-ocamlopt-save-ir/start_from_emit.ml` it's copying a single file to a new name. So semantics-wise, `src` is a list of files (or files and directories, although we don't need that at the moment) relative to the build directory, and `dst` is then either a directory or a filename, also relative to the build directory. If `src` has multiple names then `dst` must be a directory. We could go down the `cp` route of interpreting `cp src dst/` as `cp src dst/src`, but maybe it's more explicit that when copying a single item you definitely specify the full path (i.e. you must write `cp src dst/src` and `cp src dst` would be an error if `dst` existed already)? Again, we don't need it for the tests which do this at the moment, but I guess `src = "foo bar baz" dst = "foo/"` would create the directory if necessary.

I don't think the `copy` has to be done here, but I would say it's part of the original issue, yes.

For `Unix` part in ocamltest, we set the bar when ocamltest_unix was added that ocamltest needed to work without that function, which this would satisfy. For the libunix part, it doesn't seem too outrageous to require an ocamlopt-compiled ocamltest, which avoids any issue with dllunix?

shindere committed:
ocamltest: implement a copy action

shindere committed:
no-alias-deps/aliases.ml: rename b.cmi.invalid to b.cmi

This allows a simplification of the test

shindere committed:
Use the copy action in the opaque/test.ml test

shindere committed:
Use the copy action in the missing_set_of_closures test

shindere committed:
Use the copy action in the tool-ocamldep-modalias/main.ml test

shindere committed:
Use the copy action in the tool-ocamlopt-save-ir/start_from_emit.ml test

shindere commented:
Update...

David Allsopp (2021/04/12 03:41 -0700):
> I had a look at the remaining tests which still have `cp` in them:
>
> - `no-alias-deps/aliases.ml`
> 	`b.cmi.invalid` could now be renamed to `b.cmi` and committed.
> 	It couldn't be called that in the old `Makefile`-system because
> 	it would be cleaned.

Done.

> - `opaque/test.ml`
> 	The `cp` here includes items built by a previous step. This could be replaced by an action:
> ```
> **** copy
> src = "intf"
> dst = "fst"
> ```

The copy action has been implemented and is now used by the test.

> - `regression/missing_set_of_closures/missing_set_of_closures.ml`
> 	Likewise, this is copying items built by a previous step, so it
> 	could be a candidate for a new action.

Now uses copy, too.

> - `tool-ocamldep-modalias/main.ml`
> 	There are two uses of cp here: the first is to `cp lib_impl.ml lib.ml`. That could use a copy action as with the previous ones (IIUC the point is that the environment needs to be set-up without that file being captured for dependencies).
>
> 	The second use replaces `Makefile` with `Makefile.build` and
> 	`Makefile.build2` - this seems unnecessary, why don't we just
> 	use `${MAKE} -f Makefile.build` and `${MAKE} -f Makefile.build2`
> 	later on?

The second point had already been addressed, now the first one is, too.

In the same vein, shall I also add a remove action to avoid calling rm?
Should that perhaps go in its dedicated PR?

> - `tool-ocamlopt-save-ir/start_from_emit.ml`
> 	The `cp` again operates on an item in an earlier step, so it
> 	could also use a new copy action.

This also uses copy now.

> Adding a new action is probably beyond the scope of this PR (unless
> you want to), but I think it might be worth quickly fixing the
> `b.cmi.invalid` and `Makefile.build*` parts while here?

Yeah well, finally, I did it. :)

> > @@ -173,6 +173,20 @@ module Sys = struct
>      with_output_file ~bin:true dest @@ fun oc ->
>      copy_chan ic oc
>
> +  let rec copy_directory src dst =
>
> This implementation can be tripped up by symlinks - I don't think it
> matters, but we could pull `Unix.stat` into `ocamltest_unix` (as the
> machinery is there for other reasons) and reject directories
> containing symlinks?

Well not too excited about implementing this, at least not right now,
unless you insist, then I'll do :)

> > +                    error_highlighting_use2.ml \
> +                    error_highlighting_use3.ml \
> +                    error_highlighting_use4.ml"
>
> Micro-nit: these lines want an extra space!

I got it now, sorry! Should be fixed!

dra27 commented:
There's a stray blank line at line 244 in builtin_actions.ml (for in `copy_action`). I'm not entirely convinced that we needed `entry_type` rather than just:

```ocaml
  let do_copy src dst =
    let (str, f) =
      if Sys.is_directory src then
        ("directory", Sys.copy_directory)
      else
        ("file", Sys.copy_file)
```

There a `cp` to `copy` conversion missing at line 56 of tool-ocamldep-modalias/main.ml.

There are two places where `mkdir` is used (tool-command-line/test-o-one-c-file.ml and typing-shadowing-of-pervasives-submodules/redefine_largefile_top.ml) - I think these could use `subdirectories` if you tweak the definition of it so that the source directory doesn't have to exist? i.e. `subdirectories = "foo"` always creates directory `foo` in the build directory and copies directory `foo` from the test directory if it exists?

In terms of other plumbing (just grep'ing for script lines in the test headers), there are:

- mv in lib-dynlink-native/main.ml and reproducibility/cmis_on_file_system.ml (the second I think would also work with a copy, as it happens)
- rm in tool-ocamldep-modalias/main.ml
- touch tool-ocamlopt-save-ir/check_for_pack.ml and tool-ocamlopt-save-ir/start_from_emit.ml
- cmp in tool-ocamlopt-save-ir/start_from_emit.ml

At that point I think all remaining uses of `script` are either to invoke `sh` on a shell script or running programs being tested. However, I think this PR is done with the tweaks above and, if you like, the tweak to `subdirectories` (and I'm very grateful for everything that's done already!)

shindere commented:
David Allsopp (2021/04/19 03:44 -0700):
> There's a stray blank line at line 244 in builtin_actions.ml (for in `copy_action`). I'm not entirely convinced that we needed `entry_type` rather than just:
>
> ```ocaml
>   let do_copy src dst =
>     let (str, f) =
>       if Sys.is_directory src then
>         ("directory", Sys.copy_directory)
>       else
>         ("file", Sys.copy_file)
> ```

Fixed, thanks!

> There a `cp` to `copy` conversion missing at line 56 of
> tool-ocamldep-modalias/main.ml.

Fixed! Thanks!

> There are two places where `mkdir` is used
> (tool-command-line/test-o-one-c-file.ml and
> typing-shadowing-of-pervasives-submodules/redefine_largefile_top.ml) -
> I think these could use `subdirectories` if you tweak the definition
> of it so that the source directory doesn't have to exist? i.e.
> `subdirectories = "foo"` always creates directory `foo` in the build
> directory and copies directory `foo` from the test directory if it
> exists?

Probably they could, but I don't really like the idea. I'd find it much
clearer to create an mkdir action and owuldn't mind doing so, in another
PR.

> In terms of other plumbing (just grep'ing for script lines in the test headers), there are:
>
> - mv in lib-dynlink-native/main.ml and reproducibility/cmis_on_file_system.ml (the second I think would also work with a copy, as it happens)
> - rm in tool-ocamldep-modalias/main.ml
> - touch tool-ocamlopt-save-ir/check_for_pack.ml and tool-ocamlopt-save-ir/start_from_emit.ml
> - cmp in tool-ocamlopt-save-ir/start_from_emit.ml
>
> At that point I think all remaining uses of `script` are either to
> invoke `sh` on a shell script or running programs being tested.
> However, I think this PR is done with the tweaks above and, if you
> like, the tweak to `subdirectories` (and I'm very grateful for
> everything that's done already!)

Well as I said I'd prefer to implement a full mkdir action, why not in a
PR that would also address the suggestions you have mentionned above.

shindere commented:
David Allsopp (2021/04/19 03:44 -0700):
> I'm not entirely convinced that we needed `entry_type` rather than just:
>
> ```ocaml
>   let do_copy src dst =
>     let (str, f) =
>       if Sys.is_directory src then
>         ("directory", Sys.copy_directory)
>       else
>         ("file", Sys.copy_file)
> ```

I am sorry, I forgot to address this but just updated the branch with a
version that gets rid of the `entry_type` type.

One more question. In retrospect, was it such a good idea to rename the
`files` variable to `readonly_files` ? We didn't need `modifiable_files`
so far and I don't know how likely it is that we will need it one day?
It shouldn't be too had to get rid of the commit that does the renaming.
Thoughts? Opinions?

dra27 commented:
I think this is good to go, yes! For the `readonly_files`, it's up to you - I agree we've ended up in a different place from where we first expected! I think it's useful that the name hints that the files shouldn't be changed (i.e. that they might be linked), so I think it's good that they're not called `files` anymore.

I'm happy with `readonly_files` - although `additional_sources` might also work?!

shindere commented:
Thanks, David!

I prefer `files` to `sources` actually, because it's less restrictive
and we nhave a `.cmi` in the testsuite whichis not quite a `source`
file...

Thanks again for your active, warm and friendly support!

shindere committed:
Add a Changes entry

shindere commented:
Just rebased and added a Changes entry.

@dra27 what do you think about the setuplinks.shscript used in the
`tool-ocamldep-modalias` test?

Clearly it could be replaced by calls to the `copy` action but that would
make the test longer and introduce repetitions.

Do you think it's worth it?

I am willing to do the modification if you'd like me to.

dra27 commented:
My goodness, a hidden `cp`! I think it would be nicer in the header, yes... complicated test has a complicated header!

shindere committed:
Get rid of the setup-links.sh script in the tool-ocamldep-modalias test

This script was doing copies, too, which are now implemented through
ocamltest's copy action.

shindere commented:
David Allsopp (2021/04/21 03:41 -0700):
> My goodness, a hidden `cp`!

Well if you `git grep cp` in `testsuite/tests`, it's not that hidden,
actually. Amngmany false positives, you will see that there is another
one in `Makefile.build` but I guess that one has to stay.

> I think it would be nicer in the header, yes... complicated test has a
> complicated header!

Done, although I am not completely found of the change. Trying to
convince myself it's worth it because it makes things more
self-contained.

dra27 commented:
I agree it's long, but the self-contained argument feels good to me - it's one thing fewer if dispatching the testsuite artefacts to a target system to be run, for example, to have to think about.

Apart from a conflict on the `Changes` file, I think this is all good to go?

shindere commented:
David Allsopp (2021/04/21 13:44 +0000):
> I agree it's long, but the self-contained argument feels good to me -
> it's one thing fewer if dispatching the testsuite artefacts to a
> target system to be run, for example, to have to think about.

Okay, okay :)

> Apart from a conflict on the `Changes` file, I think this is all good
> to go?

Fixed, thanks. I rebased, fixed the Changes file and reordered the two
last commits.

shindere commented:
I am not sure this PR got a formal approval, but given the discussions,
the absence of conflicts and the fact that all checks passed, I took the
libert to merge it.

I will also close issue #8659 because I think it has been fully addressed
by the present PR but if not, please feel free to re-open it.

dra27 commented:
All good, @shindere, and thanks for your work on it!

shindere commented:
David Allsopp (2021/04/22 03:39 -0700):
> All good, @shindere, and thanks for your work on it!

Thanks a lot, @dra27!
